<?xml version="1.0"?>

<robot name="open_manipulator_x" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Link 1 --> 
  <link name="link1">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

  <inertial>
      <origin xyz="0.00030876154 0.00000000 -0.00012176461" />
      <mass value="0.079119962" />
    <inertia ixx="0.000012505234" ixy="0.0" ixz="-0.00000017855208"
           iyy="0.000021898364" iyz="0.0"
           izz="0.000019267361" />
  
   </inertial>   
  </link>

<!-- Joint 1 between Link 1 and Link 2 -->
  <joint name="joint1" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0.012 0.0 0.017" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit velocity="4.8" effort="1" lower="-3.1416" upper="3.1416" />
    <dynamics damping="0.1"/>
  </joint>

  <!-- Link 2 -->
  <link name="link2">
    <visual>
      <origin xyz="0 0 0.019" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 0.019" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    
    <inertial>
      <origin xyz="-0.0003018487  0.00054043684  0.047433464" />
      <mass value="0.098406837" />
      <inertia ixx="0.000034543422" ixy="-0.000000016031095" ixz="-0.00000038375155"
              iyy="0.000032689329" iyz="0.000000028511935"
              izz="0.000018850320" />
    </inertial>
  </link>

<!-- Joint 2 between Link 2 and Link 3 -->
  <joint name="joint2" type="revolute">
    <parent link="link2"/>
    <child link="link3"/>
    <origin xyz="0.0 0.0 0.0595" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit velocity="4.8" effort="1" lower="-1.5" upper="1.5" />
    <dynamics damping="0.1"/>
  </joint>

  <!-- Link 3 -->
  <link name="link3">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>

    <inertial>
      <origin xyz="0.010308393 0.00037743363 0.10170197" />
      <mass value="0.13850917" />
      <inertia ixx="0.00033055381" ixy="-0.000000097940978" ixz="-0.000038505711"
              iyy="0.00034290447" iyz="-0.0000015717516"
              izz="0.000060346498" />
    </inertial>

  </link>

<!-- Joint 3 between Link 3 and Link 4 -->  
  <joint name="joint3" type="revolute">
    <parent link="link3"/>
    <child link="link4"/>
    <origin xyz="0.024 0 0.128" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit velocity="4.8" effort="1" lower="-1.5" upper="1.4" />
    <dynamics damping="0.1"/>
  </joint>

<!-- Link 3-->  
 <link name="link4">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/link4.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>

      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/link4.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0.090909590 0.00038929816 0.00022413279" />
        <mass value="0.13274562" />
        <inertia ixx="0.000030654178" ixy="-0.0000012764155" ixz="-0.00000026874417"
                iyy="0.00024230292" iyz="0.000000011559550"
                izz="0.00025155057" />
      </inertial>
  </link> 

<!-- Joint 4 between Link 4 and Link 5 -->  
   <joint name="joint4" type="revolute">
      <parent link="link4"/>
      <child link="link5"/>
      <origin xyz="0.124 0.0 0.0" rpy="0 0 0"/>
      <axis xyz="0 1 0"/>
      <limit velocity="4.8" effort="1" lower="-1.7" upper="1.97" />
      <dynamics damping="0.1"/>
    </joint>

<!--link5 -->  
  <link name="link5">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://simulation/meshes/link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
     </visual>

    <collision>
       <origin xyz="0 0 0" rpy="0 0 0"/>
       <geometry>
        <mesh filename="package://simulation/meshes/link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      </collision>

    <inertial>
        <origin xyz="0.044206755 0.000000368399985 0.0089142216" />
        <mass value="0.14327573" />
        <inertia ixx="0.000080870749" ixy="0.0" ixz="-0.0000010157896"
                iyy="0.000075980465" iyz="0.0"
                izz="0.000093127351" />
    </inertial>

  </link> 


<!-- Joint 5 between Link 5 and GL -->  
      <joint name="gripper_left_joint" type="prismatic">
        <parent link="link5"/>
        <child link="gripper_left_link"/>
        <origin xyz="0.0817 0.021 0.0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit velocity="4.8" effort="1" lower="-0.010" upper="0.019" />
        <dynamics damping="0.1"/>
      </joint>

  <link name="gripper_left_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/gripper_left_palm.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>

      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/gripper_left_palm.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" />
        <mass value="0.001" />
        <inertia ixx="0.000001" ixy="0.0" ixz="0.0"
                iyy="0.000001" iyz="0.0"
                izz="0.000001" />
      </inertial>

  </link>

<!-- Joint 6 between Link 5 and GR -->  

    <joint name="gripper_right_joint_mimic" type="fixed">
      <parent link="gripper_right_link" />
      <child link="dummy_mimic_fix" />
    </joint>
    <link name="dummy_mimic_fix"/>

    <joint name="gripper_right_joint" type="prismatic">
      <parent link="link5"/>
      <child link="gripper_right_link"/>
      <origin xyz="0.0817 -0.021 0" rpy="0 0 0"/>
      <axis xyz="0 -1 0"/>
      <limit velocity="4.8" effort="1" lower="-0.010" upper="0.019" />
      <mimic joint="gripper_left_joint" multiplier="1"/>
      <dynamics damping="0.1"/>
    </joint>

    <link name="gripper_right_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/gripper_right_palm.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="grey"/>
      </visual>

      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://simulation/meshes/gripper_right_palm.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </collision>

     <inertial>
      <origin xyz="0 0 0" />
      <mass value="0.001" />
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0"
              iyy="0.000001" iyz="0.0"
              izz="0.000001" />
     </inertial>

    </link>

  <!-- END effector --> 
    <joint name="end_effector_joint" type="fixed">
      <parent link="link5"/>
      <child link="end_effector_link"/>
      <origin xyz="0.126 0.0 0.0" rpy="0 0 0"/>
    </joint>

    <link name="end_effector_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="0.01 0.01 0.01" />
        </geometry>
        <material name="red"/>
       </visual>
    </link>

  <!-- World -->
  <link name="world"/>

  <!-- World fixed joint-->
  <joint name="world_fixed" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="world"/>
    <child link="link1"/>
  </joint>

</robot>
